### PR TITLE
Fix eager loading of nested one-to-many relations

### LINKED ISSUE #19507: relation eager loading not working
### What steps will reproduce the problem?

There is 3 entity
  * Service - some Soft installed on VM (may be instaleld on many VMs for clustered services)
  * Comp - VM which running Service 
  * Arm - Physical PC/Laptop which runs VMs (may run many) which runs services

Tables is like
services:
  * id 

comps_in_services:
  * id
  * comps_id
  * services_id

os:
  * id
  * arms_id

arms:
  * id

so i wish to know which physical PCs involved in clustered service stability:
```
class Services extends \yii\db\ActiveRecord
{
    ...
	public function getComps()
	{
		return static::hasMany(Comps::class, ['id' => 'comps_id'])
			->viaTable('{{%comps_in_services}}', ['services_id' => 'id']);
	}
	
	public function getArms()
	{
		return static::hasMany(Arms::class, ['id' => 'arm_id'])
			->via('comps');
	}
    ...
}

class ServicesSearch extends Services
{
    ....
    public function search($params)
    {
        $query = Services::find()
        ->joinWith([
            'arms', //here is the bug comes!!!
        ]);

        $this->load($params); //filter nothing

        return new ActiveDataProvider([
            'query' => $query,
            'pagination' => false,
        ]);
}

$searchModel = new ServicesSearch();
$dataProvider = $searchModel->search([]);
var_dump($dataProvider->query->createCommand()->getRawSql());
var_dump($dataProvider->models[3]);
```

### What is the expected result?
same result on accessing arms field on services objects with or without eager loading of arms relation

### What do you get instead?
all works nice without usage of joinWith or with eagerload=false set
but if use joinWith (with eagerload enabled by default) or use "with" - some sevices loaded with empty arms field

dumped Query is same and correct in both cases
but loaded ActiveRecords objects differs (WTF???)

Without EagerLoading (columns Service,Comp,Arm): 
image

With EagerLoading:
image

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.46
| PHP version      | 7.3
| Operating system | Windows

### PR DESCRIPTION
**Pull request by spo0okie**
fix #19507

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #19507

### CODE REVIEW

File: framework/db/ActiveRelationTrait.php [hunk 1]
```diff
@@ -456,8 +456,9 @@ private function buildBuckets($models, $link, $viaModels = null, $viaQuery = nul
     private function mapVia($map, $viaMap) {
         $resultMap = [];
         foreach ($map as $key => $linkKeys) {
+            $resultMap[$key]=[];
```
Comment (bizley) [inline framework/db/ActiveRelationTrait.php:459]:
```suggestion
            $resultMap[$key] = [];
```

File: tests/data/ar/Customer.php [hunk 2]
```diff
@@ -87,6 +87,20 @@ public function getOrderItems()
         })->orderBy('id');
     }
 
+    // get OrderItems objects via orders
+    public function getOrderItems2()
```
Comment (bizley) [inline tests/data/ar/Customer.php:91]:
```suggestion
    public function getOrderItems2()
```

File: tests/data/ar/Customer.php [hunk 3]
```diff
@@ -87,6 +87,20 @@ public function getOrderItems()
...
+    {
+        return $this->hasMany(OrderItem::className(), ['order_id' => 'id'])
+            ->via('orders');
+    }
+
+    //get Items objects via Orders via OrderItems
+    public function getItems()
```
Comment (bizley) [inline tests/data/ar/Customer.php:98]:
```suggestion
    public function getItems()
```

File: tests/framework/db/ActiveRecordTest.php [hunk 4]
```diff
@@ -787,6 +787,30 @@ public function testJoinWithVia()
...
+     *
+     * @see https://github.com/yiisoft/yii2/issues/19507
+     */
+    public function testJoinWithEager()
+    {
+        $eagerCustomers=Customer::find()
+            ->joinWith([
+                'items',
+            ])
+            ->all();
```
Comment (bizley) [inline tests/framework/db/ActiveRecordTest.php:801]:
```suggestion
        $eagerCustomers = Customer::find()->joinWith(['items'])->all();
```

File: tests/framework/db/ActiveRecordTest.php [hunk 5]
```diff
@@ -787,6 +787,30 @@ public function testJoinWithVia()
...
+        $eagerItemsCount=0;
```
Comment (bizley) [inline tests/framework/db/ActiveRecordTest.php:802]:
```suggestion
        $eagerItemsCount = 0;
```

File: tests/framework/db/ActiveRecordTest.php [hunk 6]
```diff
@@ -787,6 +787,30 @@ public function testJoinWithVia()
...
+        foreach ($eagerCustomers as $customer) $eagerItemsCount+=count($customer->items);
```
Comment (bizley) [inline tests/framework/db/ActiveRecordTest.php:803]:
```suggestion
        foreach ($eagerCustomers as $customer) {
            $eagerItemsCount += count($customer->items);
        }
```

File: tests/framework/db/ActiveRecordTest.php [hunk 7]
```diff
@@ -787,6 +787,30 @@ public function testJoinWithVia()
...
+
+        $lazyCustomers=Customer::find()
+            ->all();
```
Comment (bizley) [inline tests/framework/db/ActiveRecordTest.php:806]:
```suggestion
        $lazyCustomers = Customer::find()->all();
```

File: tests/framework/db/ActiveRecordTest.php [hunk 8]
```diff
@@ -787,6 +787,30 @@ public function testJoinWithVia()
...
+        $lazyItemsCount=0;
```
Comment (bizley) [inline tests/framework/db/ActiveRecordTest.php:807]:
```suggestion
        $lazyItemsCount = 0;
```

File: tests/framework/db/ActiveRecordTest.php [hunk 9]
```diff
@@ -787,6 +787,30 @@ public function testJoinWithVia()
...
+        foreach ($lazyCustomers as $customer) $lazyItemsCount+=count($customer->items);
```
Comment (bizley) [inline tests/framework/db/ActiveRecordTest.php:808]:
```suggestion
        foreach ($lazyCustomers as $customer) {
            $lazyItemsCount += count($customer->items);
        }
```

File: tests/framework/db/ActiveRecordTest.php [hunk 10]
```diff
@@ -787,6 +787,30 @@ public function testJoinWithVia()
...
+
+        //eager loaded items
+        $this->assertEquals($eagerItemsCount, $lazyItemsCount);
```
Comment (bizley) [inline tests/framework/db/ActiveRecordTest.php:811]:
```suggestion
        $this->assertEquals($eagerItemsCount, $lazyItemsCount);
```

**Review by bizley — CHANGES_REQUESTED** [review]
Also please add Changelog line, thanks.

**Review by spo0okie — COMMENTED** [review]
ok

**Review by spo0okie — COMMENTED** [review]
sry, cant find how to approve changes =/ still see, that @bizley requested changes, while i commited everything

**Review by bizley — APPROVED** [review]

**Comment by bizley** [comment]
Just waiting for some other team member to approve it and we are good to go.

**Review by samdark — APPROVED** [review]

**Review by arogachev — APPROVED** [review]

**Comment by sigaryov** [comment]
В db/ActiveRelationTrait.php, по-моему, есть баг.

Делаю запрос:

```
->joinWith([
    'dayRealTime',
    'user.usersWorkRoles.workRole.predictionType',
])
```
Получаю ошибку:

> Undefined index: date

в функции filterByModels($models). date - это поле для связи по свойству dayRealTime.


847dd835b0

Если добавить проверку на наличие элемента в массиве, то ошибка уходит:

```
foreach ($prefixedLink as $attribute => $link) {
    $value = isset($model[$attribute]) || (is_object($model) && property_exists($model, $attribute)) ? $model[$attribute] : null;
    if ($value !== null) {
        $v[$attribute] = $model[$link];
    }
}
```

**Comment by samdark** [comment]
@sigaryov it's better to create a separate issue for it. @arogachev would you please take a look?

**Comment by arogachev** [comment]
@samdark. OK. I'll check it.

@sigaryov Thanks for reporting. Could you share the relevant data structure (you can remove sensitive data)?

**Pull request merged**